## Choose and Justify Model Performance Indicator and Choice of Algorithm

In [1]:
# need a binary classifier (R or D for each county)
# logistic regression, support vector machine are options
# for deep learning, simple neural network

In [2]:
# model performance indicator will be true positive rate and true negative rate (hit rates) vs. 
    # false negative and false positive (miss rates) because of 

## Generate Train and Test Sets

In [4]:
import pandas as pd
import numpy as np

In [2]:
# import dataset

data = pd.read_csv('features_v3.csv')
data

party  totalpop    income   service    office     drive   carpool  \
0       0.0  1.000000  0.354074  0.446203  0.654952  0.755932  0.331104   
1       0.0  0.521624  0.344999  0.417722  0.648562  0.632768  0.290970   
2       0.0  0.433955  0.337373  0.392405  0.619808  0.825989  0.371237   
3       1.0  0.400262  0.335184  0.424051  0.738019  0.795480  0.367893   
4       0.0  0.262875  0.228581  0.506329  0.769968  0.800000  0.307692   
...     ...       ...       ...       ...       ...       ...       ...   
3040    1.0  0.000017  0.335923  0.170886  0.760383  0.467797  0.324415   
3041    1.0  0.000063  0.137758  0.493671  0.252396  0.696045  0.521739   
3042    1.0  0.000018  0.192528  0.186709  0.000000  0.520904  0.521739   
3043    0.0  0.000030  0.164322  0.920886  0.523962  1.000000  0.000000   
3044    1.0  0.000000  0.421820  0.164557  0.313099  0.932203  0.000000   

       transit  workathome  meancommute  ...   SD   TN   TX   UT   VA   VT  \
0     0.110211    0.137097     0.593023  ...  0.0  0.0  0.0  0.0  0.0  0.0   
1     0.298217    0.112903     0.659884  ...  0.0  0.0  0.0  0.0  0.0  0.0   
2     0.047002    0.088710     0.540698  ...  0.0  0.0  1.0  0.0  0.0  0.0   
3     0.038898    0.158602     0.462209  ...  0.0  0.0  0.0  0.0  0.0  0.0   
4     0.089141    0.115591     0.590116  ...  0.0  0.0  0.0  0.0  0.0  0.0   
...        ...         ...          ...  ...  ...  ...  ...  ...  ...  ...   
3040  0.000000    0.607527     0.651163  ...  0.0  0.0  0.0  0.0  0.0  0.0   
3041  0.000000    0.102151     0.215116  ...  0.0  0.0  0.0  0.0  0.0  0.0   
3042  0.000000    0.534946     0.287791  ...  0.0  0.0  0.0  0.0  0.0  0.0   
3043  0.000000    0.000000     0.203488  ...  0.0  0.0  1.0  0.0  0.0  0.0   
3044  0.000000    0.096774     0.058140  ...  0.0  0.0  1.0  0.0  0.0  0.0   

       WA   WI   WV   WY  
0     0.0  0.0  0.0  0.0  
1     0.0  0.0  0.0  0.0  
2     0.0  0.0  0.0  0.0  
3     0.0  0.0  0.0  0.0  
4     0.0  0.0  0.0  0.0  
...   ...  ...  ...  ...  
3040  0.0  0.0  0.0  0.0  
3041  0.0  0.0  0.0  0.0  
3042  0.0  0.0  0.0  0.0  
3043  0.0  0.0  0.0  0.0  
3044  0.0  0.0  0.0  0.0  

[3045 rows x 64 columns]

In [3]:
from sklearn.model_selection import train_test_split

X = data[['totalpop', 'income', 'service', 'office', 'drive', 'carpool',
       'transit', 'workathome', 'meancommute', 'unemployment',
       'perc_men', 'perc_white', 'perc_private_work', 'perc_citizen', 'AL',
       'AR', 'AZ', 'CA', 'CO', 'CT', 'DC', 'DE', 'FL', 'GA', 'HI', 'IA', 'ID',
       'IL', 'IN', 'KS', 'KY', 'MA', 'MD', 'ME', 'MI', 'MN', 'MO', 'MS', 'MT',
       'NC', 'ND', 'NE', 'NH', 'NJ', 'NM', 'NV', 'NY', 'OH', 'OK', 'OR', 'PA',
       'RI', 'SC', 'SD', 'TN', 'TX', 'UT', 'VA', 'VT', 'WA', 'WI', 'WV', 'WY']]
y = data['party']

X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2)

# Logistic Regression

### Implement Logistic Regression

In [8]:
from sklearn.linear_model import LogisticRegression
model = LogisticRegression()

model.fit(X_train, y_train)

/Users/soujanyaibm.com/opt/anaconda3/lib/python3.7/site-packages/sklearn/linear_model/logistic.py:432: FutureWarning: Default solver will be changed to 'lbfgs' in 0.22. Specify a solver to silence this warning.
  FutureWarning)


LogisticRegression(C=1.0, class_weight=None, dual=False, fit_intercept=True,
                   intercept_scaling=1, l1_ratio=None, max_iter=100,
                   multi_class='warn', n_jobs=None, penalty='l2',
                   random_state=None, solver='warn', tol=0.0001, verbose=0,
                   warm_start=False)

### Predict

In [9]:
y_pred = model.predict(X_test)

### Evaluation

In [10]:
model.score(X_test, y_test)

0.9178981937602627

In [11]:
from sklearn.metrics import confusion_matrix
confusion_matrix = confusion_matrix(y_test, y_pred)
print(confusion_matrix)

[[ 62  43]
 [  7 497]]


# Neural Network

In [12]:
X.shape # 14 node input layer

(3045, 63)

In [13]:
# 24 node dense layer with sigmoid activation
# sigmoid output layer

### Implement Neural Network

In [15]:
from keras.layers import Dense
from keras.layers import Input
from keras.models import Sequential
import keras.optimizers
import keras.metrics as metrics

model = Sequential()
model.add(Dense(63, input_dim=63, activation='relu'))
model.add(Dense(20, activation='relu'))
model.add(Dense(1, activation='sigmoid'))

model.compile(optimizer='adam', loss='binary_crossentropy', metrics=[metrics.BinaryAccuracy(),metrics.AUC()])

model.fit(X_train, y_train)

Epoch 1/1
2436/2436 [==============================] - 0s 97us/step - loss: 0.4759 - binary_accuracy: 0.8190 - auc_2: 0.5704


### Predict & Evaluate

In [16]:
test_loss, test_acc = model.evaluate(X_test, y_test)[:2]

609/609 [==============================] - 0s 91us/step


In [17]:
print(test_loss, test_acc)

0.4136776596258818 0.8275862336158752


# Conclusion

Logistic regression model accuracy improved to 0.92 by using normalized column data and state dummy variables, but neural network performance didn't approve. Thus, final model definition will be Logistic Regression model using above dataset and trained on different downsampling of republican counties, since it is an unbalanced dataset.